New Tree View

bbovenzi opened the pull request:
**Problem**

A tree diagram cannot properly represent a graph. This has led to a number of issues:
- users being confused seeing multiple leaves representing the same task
- displaying task groups has never really worked well #18009, #13242
- task groups have never had a collapse/expand functionality like graph view #16112
- The long tree names and branches easily leads to a overlapping elements #17717
- It is all built in a hard-to-follow d3 setup that makes maintaining the view hard
- A recursively built tree is more computationally heavy than a flat list #17148

**Solution**

- Remove the "tree" part of tree view. Show runs and tasks but leave dependency lines to the graph view
- Allow a user to better abstract away parts of their DAGs with Task Groups and simplify their workflows
- Build in a modern technology that we can more easily maintain

**Screenshots**

<img width="1192" alt="Screen Shot 2021-10-01 at 5 16 29 PM" src="https://user-images.githubusercontent.com/4600967/135648242-717acc77-a5ab-4bd5-a560-e3199acf74a0.png">

<img width="788" alt="Screen Shot 2021-10-01 at 5 15 39 PM" src="https://user-images.githubusercontent.com/4600967/135648305-5514f9ce-f94b-4744-b5a6-279b039e2c53.png">

<img width="325" alt="Screen Shot 2021-10-01 at 5 16 48 PM" src="https://user-images.githubusercontent.com/4600967/135648325-9ae934c9-361a-4ee9-8186-8914e6064e76.png">

![Oct-01-2021 17-14-42](https://user-images.githubusercontent.com/4600967/135648403-9434f5e3-f642-445c-a843-661e58110f96.gif)

**Features and Benefits**

- Show durations of dag runs to quickly see performance changes #8931
- Task groups have summaries and can expand/collapse just like the graph view
- Task groups default to collapsed for an easier quick view but user changes are saved via localstorage
- Grid lines and hover effects to see which task you are inspecting
- Built in React to more easily maintain and can bring over to an eventual new Airflow UI in AIP-38
- Ready to add new features in the roadmap like Dag Versioning or Task Mapping
- Include a play icon to better indicate manual dag runs (dark borders weren't very intuitive)
- Testing of React components

Subsequent PRs:

- Rename Tree view (Grid?) and update all documentation
- Filter over the state legend to highlight tasks and dag runs like graph view
- Update default colors and better support color-blind friendly modes
- More elegantly handle outliers in the Dag Run Duration chart (if it normally takes 5 minutes but then a run takes 5 hours the graph won't look great)
- Update the Task Instance and Dag Run Action Modals too
- More dynamic scrolling around the 100s of dag runs instead of the clunky "Update" button
- Improve render times with Lazy Loading #13066

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

bbovenzi committed:
first pass at new matrix view

setup jsx inside existing project

React task group tree view

refactor merge conflicts

flexbox and tooltip styling

full tooltips

column highlighing and saving open groups

remove active run logic

auto refresh

determine state, start_date and end_date for a taskgroup based on the children

remove join nodes

Add task instances to output

improve return value for tree data

Add group tis

Filter data by dag_run

Revert "Filter data by dag_run"

This reverts commit badf509c99acc65b6a74e775414ecd90518a6dba.

update group instance summaries

fix summary dates, active run check, and tooltips

add task instance modal

add dag run modal

use colors from settings

refactor jsx folder structure and some linting

eslint jsx filesand no autorefresh when paused

improve dag run axes

variosu UI improvements

clean up duration tick code

background color for groups

small graph improvements

basic overflow handling

improve dates

add horizontal hover background

remove all /ui/ changes

use auto_refresh_interval

updated status and axes

use run_id to filter

handle loading and empty views

fix group runs order and remove group pointer

fix reversed group instance summaries

update hover/group colors and add vertical tick marks

use op.extralinks instead of get_task

remove redundant dayjs package

update babel packages

- trying to see if babel upgrades will fix CI Build issue with parsing jsx

Revert "update babel packages"

This reverts commit 1e94af84ad23a224c8d1650e9f5826810ffe9100.

better names in views, parse jsx in webpack, remove unnecessary test

fix typing import order in utils

uranusjr commented on airflow/www/views.py:
```diff
@@ -219,6 +219,82 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
+    Create a nested dict representation of this TaskGroup and its children used to construct
+    the Graph.
+    """
+    if isinstance(task_group, BaseOperator):
```
The variable should probably not be named `task_group` 😛

ephraimbuddy replied:
I can't think of a better name 🤔 Do you have suggestion on the name? I used `task_object` but it didn't go down well with me.

uranusjr replied:
`task_item_or_group`? (And we can assign it to new names after the `isinstance()` check)

uranusjr commented on airflow/www/views.py:
```diff
@@ -219,6 +219,82 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
+            'id': task_group.task_id,
+            'instances': [wwwutils.encode_ti(ti) for ti in tis if ti.task_id == task_group.task_id],
+            'label': task_group.label,
+            'extra_links': dag.get_task(task_group.task_id).extra_links,
```
Is this needed? `DAG.get_task()` also returns `BaseOperator`.

uranusjr commented on airflow/www/views.py:
```diff
@@ -219,6 +219,82 @@ def get_date_time_num_runs_dag_runs_form_data(www_request, session, dag):
+            'skipped',
+        ]
+
+        child_instances = [child['instances'] for child in children if child.get('instances')]
```
```suggestion
        child_instances = [child['instances'] for child in children if 'instances' in child]
```

ephraimbuddy replied:
Will fix this and some other reviews, also see if testing can be possible

mik-laj commented on airflow/www/package.json:
```diff
@@ -56,19 +60,29 @@
     "nvd3": "^1.8.6",
+    "react": "^17.0.2",
+    "react-dom": "^17.0.2",
+    "react-icons": "^4.2.0",
```
Should we add this library or related to [`airflow/licences/`](https://github.com/apache/airflow/tree/main/licenses)?

mik-laj replied:
I am not sure about the licenses for the following libraries, but I have not checked all licenses and all libraries.
```
 ast-types-flow 0.0.7 released 6 years ago
 damerau-levenshtein 1.0.7 released 5 months ago
 hoist-non-react-statics 3.3.2 released 2 years ago
 type-fest 1.4.0 released 2 months ago
```

ashb replied:
I checked these libs and they are MIT, BSD (2 or 3 caluse) and ISC which are all category A/allowed for ASF.

ashb replied:
Oh, we'll need to include the license for _anything_ that webpack pulls in and thus we ship directly as part of the our artefacts. That is... tedius.

@bbovenzi Something like https://github.com/codepunkt/webpack-license-plugin might be what we need here - us trying to maintain it manually is just never going to work.

ashb replied:
(I just picked that somewhat at random -- it may not actually do what we want or be the best option)

bbovenzi replied:
I updated the PR to use the above library in our `webpack.config.js` to auto-generate a `licenses/LICENSES-ui.txt` file when running `yarn run build`. The file is gitignored but will be included in any builds during `setup.py`

mik-laj commented on airflow/www/static/js/tree/DagRunBar.jsx:
```diff
@@ -0,0 +1,84 @@
+    {index < totalRuns - 4 && index % 10 === 0 && (
+    <VStack position="absolute" top="0" left="-23px" spacing={0}>
+      <Text fontSize={10} color="gray.400" whiteSpace="nowrap" transform="rotate(-30deg) translateX(32px)" mt="-23px !important">
+        {moment.utc(run.executionDate).format('MMM DD, HH:mm')}
```
Why do we sometimes use the `moment.js` library and other times the day.js library?

mik-laj commented on airflow/www/static/js/tree/InstanceTooltip.jsx:
```diff
@@ -0,0 +1,155 @@
+      <Text>
+        Duration:
+        {' '}
+        {dayjs.duration(instanceDuration).days() > 0 && `${dayjs.duration(instanceDuration).days()}d `}
```
Should we extract this fragment to `formatDuration` function similar to `formatDateTime`?

mik-laj commented on airflow/www/static/js/tree/useTreeData.js:
```diff
@@ -0,0 +1,79 @@
+    const resp = await fetch(`${treeDataUrl}?dag_id=${dagId}&num_runs=${numRuns}&root=${urlRoot}`);
+    let newData = await resp.json();
+    newData = camelcaseKeys(newData, { deep: true });
+    if (JSON.stringify(newData) !== JSON.stringify(data)) {
```
Are we sure that all elements will always have the same order?

bbovenzi replied:
I guess not, but if the order does change its not a bad idea to rerender the tree anyway. 
I'm open to suggestions on how we should order tasks in the webserver.

uranusjr replied:
Python dicts are insertion-ordered (from 3.6 up at least), so this should be mostly stable. The most likely source of issues would be the first refresh, because the initial `treeData` is from the HTML, but subsequent runs use the JSON-based view, and those are more likely to have inconsistencies if we're not careful. I kind of wonder if we should get rid of the initial `treeData` and always render from the JSON view.

The other significant possibility of inconsistencies is SQL natural ordering, but those can be caught incrementally if needed.

SamWheating commented on airflow/www/views.py:
```diff
@@ -2283,9 +2260,10 @@ def tree(self, session=None):
+        dag_run_dates = {dr.execution_date: alchemy_to_dict(dr) for dr in dag_runs}
 
-        max_date = max(dag_runs.keys(), default=None)
+        max_date = max(dag_run_dates.keys(), default=None)
```
```suggestion
        max_date = max(dag_run_dates, default=None)
```

I think that `max(dict)` will return the max key by default.

bbovenzi committed:
add 3rd party licenses

- Add licenses for npm modules added in `dependencies` of `package.json`

bbovenzi committed:
add licenses to setup.cfg file

bbovenzi committed:
UI clean up and bug fixes

- Fix autorefresh
- remove hyphen in run y-axis
- increase boxes from 10px to 12px
- add 100ms delay to tooltips
- handle long task names
- handle missing task instances

ephraimbuddy committed:
Update airflow/www/views.py

Co-authored-by: Sam Wheating <samwheating@gmail.com>

ephraimbuddy committed:
use task_item_or_group instead of task_group in task_group_to_tree

bbovenzi committed:
Include dag-file tooltip for task groups

ryanahamilton commented on airflow/www/static/js/tree/DagRunTooltip.jsx:
```diff
@@ -0,0 +1,102 @@
+  },
+}) => (
+  <Box fontSize={12}>
+    <br />
```
Are the `<br />`s at the top and bottom necessary? If you do want extra spacing, maybe add `py={4}` to the parent `Box` instead?

ryanahamilton commented on airflow/www/static/js/tree/Tree.jsx:
```diff
@@ -0,0 +1,202 @@
+          <Flex
+            as={isGroup ? 'button' : 'div'}
+            onClick={() => isGroup && onToggle()}
+            color={(level > 4 || (isHover && level > 3)) && 'white'}
```
Wondering if you could avoid the `isHover` state with something like this?
```suggestion
          role="group"
          _hover={{ backgroundColor: 'rgba(113, 128, 150, 0.1'}}
        >
          <Flex
            as={isGroup ? 'button' : 'div'}
            onClick={() => isGroup && onToggle()}
            color={level > 4 ? 'white' : undefined}
            _groupHover={level > 3 && {
              color: 'white',
            }}
```

bbovenzi committed:
jsx code improvements

bbovenzi committed:
Update to include overflow scrolling

- Switch from FlexBox to Table elements to support horizontally scrolling of dag runs/task instances

bbovenzi committed:
fix task props, use px for sizes, update documentation

bbovenzi committed:
add ui testing for react components

ashb commented on airflow/www/static/js/tree/InstanceTooltip.jsx:
```diff
@@ -0,0 +1,145 @@
+      ['scheduled', 0],
+      ['skipped', 0],
+      ['no_status', 0],
+    ]);
```
We need "sensing" in this list I think -- or to decide that 2.3 is where we are going to remove the experimental SmartSensor feature.

bbovenzi replied:
Added 'sensing' for now.

ashb commented on airflow/www/static/js/tree/utils.js:
```diff
@@ -0,0 +1,31 @@
+  const days = moment.duration(duration).days();
+  // .as('milliseconds') is necessary for .format() to work correctly
+  return `${days > 0 ? `${days}d` : ''}${moment.utc(moment.duration(duration).as('milliseconds')).format('HH:mm:ss')}`;
+};
```
Rather than a new generic `utils.js` file (which I worry will end up an unstructured dumping ground) these Duration functions would be equally at home in the existing `datetime_utls` I think.

ashb commented on airflow/www/static/js/tree/utils.js:
```diff
@@ -0,0 +1,31 @@
+export const formatDuration = (duration) => {
+  const days = moment.duration(duration).days();
+  // .as('milliseconds') is necessary for .format() to work correctly
+  return `${days > 0 ? `${days}d` : ''}${moment.utc(moment.duration(duration).as('milliseconds')).format('HH:mm:ss')}`;
```
No need to call `moment.duration()` twice.

```suggestion
  duration = moment.duration(duration)
  const days = duration.days();
  // .as('milliseconds') is necessary for .format() to work correctly
  return `${days > 0 ? `${days}d` : ''}${moment.utc(duration.as('milliseconds')).format('HH:mm:ss')}`;
```

ashb commented on airflow/www/templates/airflow/tree.html:
```diff
@@ -66,12 +54,7 @@
-        {{ op.task_type }}
-      </span>{% endfor %}
-    </div>
+    <div></div>
```
What's this empty div needed for in the HTML?

ashb reviewed (commented):
A few small comments. Looking good though!

bbovenzi committed:
consolidate utils, empty data tests, remove extra div

bbovenzi committed:
update licenses

bbovenzi committed:
remove now extraneous license checks

bbovenzi committed:
update setup.cfg

bbovenzi committed:
remove pre-commit added line

ryanahamilton commented on airflow/www/webpack.config.js:
```diff
@@ -35,6 +36,22 @@ const JS_DIR = path.resolve(__dirname, './static/js');
+// Convert licenses json into a standard format for LICENSES.txt
+const formatLicenses = (packages) => {
+  let text = `Apache Airflow
+Copyright 2016-2021 The Apache Software Foundation
```
Could we interpolate the `2021` to save us the hassle of modifying two months from now?

bbovenzi replied:
I copied this from `NOTICE`. Is there a way to interpolate it there too? or is `2016-present` allowed?

potiuk replied:
I believe it's not allowed.

My interpretation is that you have to state what was the year your contributions were made and it should not be 
  a) future, because for example it would limit you so that you could not change licencing scheme until it expires 
  b) `present` verbatim - because it is meaningless in the context of "writer". The `present` only makes sense in the context of "reader" because you do not keep in the file the date when "present" was put. However, the licence file needs the context of "writer", so you need to specify when was the time the licence was last "expected" to be granted by those who have the rights to grant the licence.

We usually perform a bulk update of all licence dates at the beginning of the year. This is also clear expression of intent "yes we want to licence our code under the ASF licence also in 2022". So do not worry about it now.

ryanahamilton reviewed (approved):
Really amazing work @bbovenzi! 🔥🔥🔥
